### DISCUSSION TITLE
How to handle failed write and how should `terminate` be used?

### DISCUSSION THREAD

**Discussion by yun-li** [foxglove/mcap#1076] [General]
I have an C++ MCAP writer who writes mcap message. After a message is written, it needs to check if the write is successful. If not, we need to call `terminate`, can we write more messages after the terminate call?

For example, can we do think like

```
for (int i = 0; i < 10; i++) {
    const auto res = writer.write(message[i])
    if (!res.ok()) {
      std::cerr << "Failed to write message: " << res.message << "\n";
      writer.terminate(); // do not close the writer, and continue writing the file in the next for-loop
    }
}
```
Does that work? Or do we need to call `writer.close()` after `writer.terminate()`? 
Can we continue writing message to the same file by reopen the file after it terminates? Or do we need to open a new file?

I know according to this example https://github.com/foxglove/mcap/blob/main/cpp/examples/protobuf/writer.cpp#L155
When the write failed, 
```
    const auto res = writer.write(msg);
    if (!res.ok()) {
      std::cerr << "Failed to write message: " << res.message << "\n";
      writer.terminate();
      writer.close();
      std::ignore = std::remove(outputFilename);
      return 1;
    }
```
It closes every thing and terminate the program, which is not what we want.

Thanks!

**Comment by james-rms** [comment]
The `IWritable` interface does not communicate issues that happen in the filesystem write() call back to the caller. The default `FileWriter` implementation that comes with the MCAP C++ library simply asserts that the write succeeded, which will cause your program to terminate early in debug mode, or fail silently in release mode if there is a write error.

The merits of this approach are debatable, but the bottom line today is that if you want to handle file write errors, you need to implement your own `IWritable` which communicates write errors to the application through a side effect or a callback.

The purpose of the `Writer::write()` return value `Status` is to communicate errors around the way the `Writer` is being used - for example, if you write a message before calling open() or before writing the schema for that message. Errors like this are not recoverable - the right thing is usually to close the file, terminate the application, then later change the code that used the `Writer` API in a way which caused an error.

**Reply by yun-li** [reply]
OK. Thanks!

Do you recommend using `try-catch` block to catch the error, and then reopen the file for usecases where application failure is not acceptable? 

With the write error, would that MCAP file still decodable after calling `terminate` and `close`?

Thanks!
